lib/utils/database.dart

```diff
@@ -1,36 +1,55 @@
-import '../models/message.dart';
-import '../models/device.dart';
+import 'package:chirp_nets/models/user.dart';
+import 'package:chirp_nets/models/conversation.dart';
+import 'package:chirp_nets/models/contact.dart';
+import 'package:chirp_nets/models/message.dart';
+import 'package:chirp_nets/models/device.dart';
 import 'package:path/path.dart';
 import 'package:sqflite/sqflite.dart';
-import 'dart:async';
+
+final List<String> migrations = [
+  "CREATE TABLE users(id INTEGER PRIMARY KEY, name TEXT);",
+  "CREATE TABLE device(id INTEGER PRIMARY KEY, userId INTEGER, deviceRSSI TEXT, FOREIGN KEY(userId) REFERENCES users(id));",
+  "CREATE TABLE conversations(id INTEGER PRIMARY KEY, userId INTEGER, name TEXT, FOREIGN KEY(userId) REFERENCES users(id));",
+  "CREATE TABLE messages(id INTEGER PRIMARY KEY, conversationId INTEGER, message TEXT, createdAt TEXT, FOREIGN KEY(conversationId) REFERENCES conversations(id));",
+  "CREATE TABLE contact(id INTEGER PRIMARY KEY, conversationId INTEGER, name TEXT, deviceId TEXT, FOREIGN KEY(conversationId) REFERENCES conversations(id));",
+  "ALTER TABLE device ADD COLUMN name;",
+  "ALTER TABLE messages ADD COLUMN createdBy INTEGER; ALTER TABLE messages add FOREIGN KEY(createdBy) REFERENCES users(id)",
+  "ALTER TABLE contact ADD COLUMN userId INTEGER; ALTER TABLE contact add FOREIGN KEY(userId) REFERENCES users(id)",
+];
 
 Future<Database> getDatabase() async {
   final Future<Database> database = openDatabase(
     join(await getDatabasesPath(), 'chirpnets.db'),
     onCreate: (db, version) {
-      return db.execute(
-        """
-          CREATE TABLE users(id INTEGER PRIMARY KEY, name TEXT); 
-          CREATE TABLE device(id INTEGER PRIMARY KEY, userId INTEGER, deviceRSSI TEXT, FOREIGN KEY(userId) REFERENCES users(id));
-          CREATE TABLE conversations(id INTEGER PRIMARY KEY, userId INTEGER, name TEXT, FOREIGN KEY(userId) REFERENCES users(id));
-          CREATE TABLE messages(id INTEGER PRIMARY KEY, conversationId INTEGER, message TEXT, createdAt TEXT, FOREIGN KEY(conversationId) REFERENCES conversations(id));
-          CREATE TABLE contact(id INTEGER PRIMARY KEY, conversationId INTEGER, name TEXT, FOREIGN KEY(conversationId) REFERENCES conversations(id));
-        """,
-      );
+      for (var script in migrations) {
+        db.execute(script);
+      }
+    },
+    onUpgrade: (db, oldVersion, newVersion) async {
+      var batch = db.batch();
+      print(oldVersion);
+      print(newVersion);
+      for (var i = oldVersion - 1; i == newVersion - 1; i++) {
+        batch.execute(migrations[i]);
+      }
+      await batch.commit();
     },
-    version: 1,
+    version: migrations.length,
   );
   return database;
 }
 
+void close() async {
+  var db = await database;
+  await db.close();
+}
+
 final Future<Database> database = getDatabase();
 
-Future<void> create({String table, dynamic object}) async {
+Future<int> create({String table, dynamic object}) async {
```

Comment:
This Method is async, but has await on every line. I think that makes the async redundant as you are doing everything sync

Reply:
The method is async so when this is called the method that called it continues. The stuff in this needs to be synchronous because you can't insert into a db if you dont have it

Reply:
The method is async so when this is called the method that called it continues. The stuff in this needs to be synchronous because you can't insert into a db if you dont have it